dbms/src/Processors/QueryPipeline.cpp

```diff
@@ -48,36 +47,41 @@ void QueryPipeline::checkSource(const ProcessorPtr & source, bool can_have_total
                         toString(source->getOutputs().size()) + " outputs.", ErrorCodes::LOGICAL_ERROR);
 }
 
-void QueryPipeline::init(Processors sources)
+void QueryPipeline::init(Pipe pipe)
+{
+    Pipes pipes;
+    pipes.emplace_back(std::move(pipe));
+    init(std::move(pipes));
+}
+
+void QueryPipeline::init(Pipes pipes)
 {
     if (initialized())
         throw Exception("Pipeline has already been initialized.", ErrorCodes::LOGICAL_ERROR);
 
-    if (sources.empty())
-        throw Exception("Can't initialize pipeline with empty source list.", ErrorCodes::LOGICAL_ERROR);
+    if (pipes.empty())
+        throw Exception("Can't initialize pipeline with empty pipes list.", ErrorCodes::LOGICAL_ERROR);
 
     std::vector<OutputPort *> totals;
 
-    for (auto & source : sources)
+    for (auto & pipe : pipes)
     {
-        checkSource(source, true);
-
-        auto & header = source->getOutputs().front().getHeader();
+        auto & header = pipe.getHeader();
 
         if (current_header)
             assertBlocksHaveEqualStructure(current_header, header, "QueryPipeline");
```

Comment:
Previously we did this check in debug mode only https://github.com/clickhouse/ClickHouse/blob/master/dbms/src/DataStreams/IBlockInputStream.cpp#L94.

Reply:
In `IBlockInputStream` it is done once per block. But in `QueryPipeline` only once while building. It's almost free check.